superduperdb/base/document.py

```diff
@@ -204,6 +204,7 @@ def _encode(
         return out
     # ruff: noqa: E501
     if isinstance(r, Leaf) and not isinstance(r, leaf_types_to_keep):  # type: ignore[arg-type]
+        # TODO: (not leaf_types_to_keep or isinstance(r, leaf_types_to_keep)) ?
```

Comment:
Is that a bug?

Reply:
No. The idea is that sometimes, you don't want to encode everything.

Reply:
Maybe need to change the codition to `isinstance(r, Leaf) and (not leaf_types_to_keep or isinstance(r, leaf_types_to_keep))` 

if leaf_types_to_keep is not empty tuple then the `isinstance(r, Leaf) and not isinstance(r, leaf_types_to_keep)` will return `false`

```
In [1]: from superduperdb.ext.torch.model import TorchModel
   ...: from superduperdb import Document
   ...: import torch
   ...: model = TorchModel(object=torch.nn.Linear(10, 10), identifier='model')
   ...: from superduperdb.ext.sklearn.model import Estimator
   ...: Document({"model": model}).encode(leaf_types_to_keep=(Estimator, ))
   ...:
Out[1]:
{'model': {'_content': {'cls': 'TorchModel',
   'module': 'superduperdb.ext.torch.model',
   'type_id': 'model',
   'identifier': 'model',
   'version': None,
   'hidden': False,
   'leaf_type': 'component'}}}

```


```
In [3]: from superduperdb.ext.torch.model import TorchModel
   ...: from superduperdb import Document
   ...: import torch
   ...: model = TorchModel(object=torch.nn.Linear(10, 10), identifier='model')
   ...: from superduperdb.ext.sklearn.model import Estimator
   ...: Document({"model": model}).encode(leaf_types_to_keep=(TorchModel, ))
Out[3]: {'model': TorchModel(identifier='model', datatype=None, output_schema=None, flatten=False, preprocess=None, postprocess=None, collate_fn=None, batch_predict=False, takes_context=False, metrics=(), model_update_kwargs={}, validation_sets=None, predict_X=None, predict_select=None, predict_max_chunk_size=None, predict_kwargs=None, object=Linear(in_features=10, out_features=10, bias=True), model_to_device_method='to', metric_values={}, predict_method=None, device='cpu', preferred_devices=('cuda', 'mps', 'cpu'), training_configuration=None, train_X=None, train_y=None, train_select=None, num_directions=2, optimizer_state=None, forward_method='__call__', train_forward_method='__call__')}

```

Reply:
> No. The idea is that sometimes, you don't want to encode everything.

Is this expected? This parameter means that the specific object will not be encoded during encoding, right?

```
In [1]: from superduperdb.ext.torch.model import TorchModel
   ...: from superduperdb import Document
   ...: import torch
   ...: model = TorchModel(object=torch.nn.Linear(10, 10), identifier='model')
   ...: from superduperdb.ext.sklearn.model import Estimator
   ...: Document({"model": model}).encode(leaf_types_to_keep=(Estimator, ))
   ...:
Out[1]:
{'model': {'_content': {'cls': 'TorchModel',
   'module': 'superduperdb.ext.torch.model',
   'type_id': 'model',
   'identifier': 'model',
   'version': None,
   'hidden': False,
   'leaf_type': 'component'}}}

```


```
In [3]: from superduperdb.ext.torch.model import TorchModel
   ...: from superduperdb import Document
   ...: import torch
   ...: model = TorchModel(object=torch.nn.Linear(10, 10), identifier='model')
   ...: from superduperdb.ext.sklearn.model import Estimator
   ...: Document({"model": model}).encode(leaf_types_to_keep=(TorchModel, ))
Out[3]: {'model': TorchModel(identifier='model', datatype=None, output_schema=None, flatten=False, preprocess=None, postprocess=None, collate_fn=None, batch_predict=False, takes_context=False, metrics=(), model_update_kwargs={}, validation_sets=None, predict_X=None, predict_select=None, predict_max_chunk_size=None, predict_kwargs=None, object=Linear(in_features=10, out_features=10, bias=True), model_to_device_method='to', metric_values={}, predict_method=None, device='cpu', preferred_devices=('cuda', 'mps', 'cpu'), training_configuration=None, train_X=None, train_y=None, train_select=None, num_directions=2, optimizer_state=None, forward_method='__call__', train_forward_method='__call__')}

```

Reply:
Yes this is sometimes useful.